src/middlewared/middlewared/plugins/smb.py

```diff
@@ -72,6 +72,130 @@ def key_cp(encoding):
+            return {'status': True, 'message': None}
+
+        for group in member_list.splitlines():
+            group = group.rstrip()
+            if group == sid:
+                return {'status': True, 'message': 'SID_IS_PRESENT'}
+            if group.rsplit('-', 1)[-1] != "512" and group != sid:
+                self.logger.debug(f"Removing {group} from local admins group.")
+                rem = await Popen(
+                    ['/usr/local/bin/net', 'groupmap', 'delmem', 'S-1-5-32-544', group],
+                    stdout=subprocess.PIPE, stderr=subprocess.PIPE
+                )
+                remout = (await rem.communicate())[1].decode()
+                if remout:
+                    ret['message'] = f"failed to remove {group} from local admins group"
+                    return ret
+
+        return {'status': True, 'message': None}
+
+    @private
+    async def wbinfo_gidtosid(self, gid):
+        ret = {
+            'status': False,
+            'message': None,
+        }
+        if not gid:
+            return ret
+        proc = await Popen(
+            ['/usr/local/bin/wbinfo', '--gid-to-sid', f"{gid}"],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        output = (await proc.communicate())
+        if output[1]:
+            ret['message'] = output[1].decode()
+        else:
+            ret['status'] = True
+            ret['message'] = output[0].decode().rstrip()
+
+        return ret
+
+    """
+        Add a local or directory service group to BUILTIN\\Administrators (S-1-5-32-544)
+        Members of this group have elevated privileges to the samba server (ability to
+        take ownership of Files, override ACLs, view and modify user quotas, and administer
+        the server via the Computer Management MMC Snap-In. Unfortuantely, group membership
+        must be managed via "net groupmap listmem|addmem|delmem", which requires that
+        winbind be running when the commands are executed. In this situation net command
+        will fail with WBC_ERR_WINBIND_NOT_AVAILABLE. If this error message is returned, then
+        flag for a deferred command retry when service starts.
+
+        @param-in (admin_group): This is the group to add to BUILTIN\\Administrators. If unset, then
+            look up the value in the config db.
+        @param-in (check_deferred): If this is True, then only perform the group mapping if this has
+            been flagged as in need of deferred setup (i.e. samba wasn't running when it was initially
+            called). This is to avoid unecessarily calling during service start.
+
+        @param-out dict 'status'- True or False, 'message'- error messages in case of failure.
+    """
+    @accepts(
+        Str('admin_group'),
+        Bool('check_deferred'),
+    )
+    async def add_admin_group(self, admin_group=None, check_deferred=False):
+        ret = {
+            'status': False,
+            'message': None,
+        }
```

Comment:
Just as an idea, sometimes on python its better to use exception for situations like this, it may read better instead of passing status with message around.

Not need to change it if you dont think its useful/worthwhile.

Reply:
I will do this.